### PR TITLE
Infrared Emitters: THE COMEBACK! ("Five Years Later" edition)

### PR DESCRIPTION
**Pull request by DeityLink**
image
emittersinfraredhaha

:cl:
* rscadd: After having disappeared for 5 years, Infrared Emitters are back in the assembly family! Available from vending machines (Vendomat, Toximate 3000), cargo (Plasma Assembly kit), and the autolathe. 
* bugfix: As a side effect, getting hit by stuff that deals 0 damage no longer brings up the damage overlay that makes you think you got hurt while you didn't.

### CODE REVIEW

**Comment by Lokiamis** [comment]
Always wondered why these were removed

**Comment by Konater** [comment]
they only had one direction

**Comment by DeityLink** [comment]
@Lokiamis too many reasons

@Konater among other things. These things actually pre-date beams, but whoever ported them to use nu-emitter beams didn't finish the job, leaving them buggy, laggy, and overall not even triggering at all.

**Comment by SonixApache** [comment]
fond memories of setting up cargo desk bells, good shit deity

File: code/modules/assembly/infrared.dm [hunk 1]
```diff
@@ -141,70 +171,143 @@
...
+/obj/effect/beam/infrared/get_damage()
+	return 0
+
+/obj/effect/beam/infrared/update_icon()
+	puffed = 0
+	if (!master)
+		invisibility = 101
+	else if (visible)
+		invisibility = 0
+		alpha = 255
```
Comment (jknpj) [inline code/modules/assembly/infrared.dm]:
```suggestion
		alpha = OPAQUE
```

File: code/modules/assembly/infrared.dm [hunk 2]
```diff
@@ -141,70 +171,143 @@
...
+	else
+		invisibility = INVISIBILITY_LEVEL_ONE
+		alpha = 128
```
Comment (jknpj) [inline code/modules/assembly/infrared.dm]:
```suggestion
		alpha = SEMI_TRANSPARENT
```

File: code/modules/assembly/infrared.dm [hunk 3]
```diff
@@ -141,70 +171,143 @@
...
+	children = list()
+	sources = list()
+
+/obj/effect/beam/infrared/get_damage()
+	return 0
+
+/obj/effect/beam/infrared/update_icon()
+	puffed = 0
+	if (!master)
+		invisibility = 101
```
Comment (jknpj) [inline code/modules/assembly/infrared.dm]:
```suggestion
		invisibility = INVISIBILITY_MAXIMUM
```

File: code/modules/assembly/infrared.dm [hunk 4]
```diff
@@ -141,70 +171,143 @@
...
+	else if (visible)
+		invisibility = 0
```
Comment (jknpj) [inline code/modules/assembly/infrared.dm:242]: There's no define for this, sadly :pensive:

File: code/modules/assembly/infrared.dm [hunk 5]
```diff
@@ -141,70 +171,143 @@
...
+	B.visible=visible
+	B.assembly=assembly
+	return B
 
 /obj/effect/beam/infrared/proc/set_visible(v)
 	visible = v
+	if (master)
+		if (visible)
+			invisibility = 0
+			alpha = 255
```
Comment (jknpj) [inline code/modules/assembly/infrared.dm]:
```suggestion
			alpha = OPAQUE
```

File: code/modules/assembly/infrared.dm [hunk 6]
```diff
@@ -141,70 +171,143 @@
...
+		else
+			invisibility = INVISIBILITY_LEVEL_ONE
+			alpha = 128
```
Comment (jknpj) [inline code/modules/assembly/infrared.dm]:
```suggestion
			alpha = SEMI_TRANSPARENT
```

File: code/modules/assembly/infrared.dm (hunk 4 above)
Comment (DeityLink) [reply code/modules/assembly/infrared.dm:242]: 😔

File: code/modules/assembly/infrared.dm [hunk 7]
```diff
@@ -10,101 +10,132 @@
...
 		if(beam)
 			returnToPool(beam)
+			beam = null
 		processing_objects.Remove(src)
 	update_icon()
 	return secured
 
-
 /obj/item/device/assembly/infra/update_icon()
 	overlays.len = 0
```
Comment (jknpj) [inline code/modules/assembly/infrared.dm:73]:
```suggestion
	overlays.Cut()
```

File: code/modules/assembly/infrared.dm [hunk 8]
```diff
@@ -10,101 +10,132 @@
...
 	attached_overlays = list()
-	if(on)
-		attached_overlays += "infrared_on"
-		overlays += image(icon = icon, icon_state = "infrared_on")
+	attached_overlays["infrared_aim"] = image(icon = icon, icon_state = "infrared_aim", dir = dir)
+	overlays += attached_overlays["infrared_aim"]
+
+	if(on && visible)
+		var/image/I = image(icon = icon, icon_state = "infrared_on", dir = dir)
```
Comment (jknpj) [inline code/modules/assembly/infrared.dm:79]:
```suggestion
		var/image/I = image(icon = src.icon, icon_state = "infrared_on", dir = src.dir)
```
THIS TIME src is fine.

File: code/modules/assembly/infrared.dm [hunk 9]
```diff
@@ -10,101 +10,132 @@
...
 	return secured
 
-
 /obj/item/device/assembly/infra/update_icon()
 	overlays.len = 0
 	attached_overlays = list()
-	if(on)
-		attached_overlays += "infrared_on"
-		overlays += image(icon = icon, icon_state = "infrared_on")
+	attached_overlays["infrared_aim"] = image(icon = icon, icon_state = "infrared_aim", dir = dir)
```
Comment (jknpj) [inline code/modules/assembly/infrared.dm:75]:
```suggestion
	attached_overlays["infrared_aim"] = image(icon = src.icon, icon_state = "infrared_aim", dir = src.dir)
```
THIS TIME src is fine.

File: code/modules/assembly/infrared.dm [hunk 10]
```diff
@@ -10,101 +10,132 @@
...
 	return 1
 
-
 /obj/item/device/assembly/infra/toggle_secure()
 	secured = !secured
-	if(secured)
-		processing_objects.Add(src)
-	else
+	if(!secured)
 		on = 0
```
Comment (jknpj) [inline code/modules/assembly/infrared.dm]:
```suggestion
		on = FALSE
```

File: code/modules/assembly/infrared.dm (hunk 7 above)
Comment (DeityLink) [reply code/modules/assembly/infrared.dm:73]: can I not do that? I've been conditioned by @d3athrow to use len = 0
https://github.com/vgstation-coders/vgstation13/commit/72fde47f91f7d4fed3a34f0175a0ff8fdab0b043#diff-f8a195145812f9e4ace5ed0ff4643950

File: code/modules/assembly/infrared.dm [hunk 11]
```diff
@@ -10,101 +10,132 @@
...
 	if(!..())
 		return 0//Cooldown check
 	on = !on
+	var/turf/T = get_turf(src)
+	if (T)
+		playsound(T,'sound/misc/click.ogg',30,0,-5)
+
+	if (on)
+		processing_objects.Add(src)
+		playsound(T,'sound/weapons/egun_toggle_laser.ogg',70,0,-5)
```
Comment (jknpj) [inline code/modules/assembly/infrared.dm:50]: @ShiftyRail is going to *freak*

File: code/modules/assembly/infrared.dm (hunk 7 above)
Comment (jknpj) [reply code/modules/assembly/infrared.dm:73]: Sure

**Pull request merged**